internal/services/tesla_api_service.go

```diff
@@ -0,0 +1,129 @@
+package services
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/DIMO-Network/devices-api/internal/config"
+	"github.com/rs/zerolog"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+//go:generate mockgen -source tesla_api_service.go -destination mocks/tesla_api_service_mock.go
+type TeslaAPIService interface {
+	CompleteTeslaAuthCodeExchange(authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error)
+	GetVehicles(token, region string) ([]TeslaVehicle, error)
+}
+
+type teslaAPIService struct {
+	Settings   *config.Settings
+	HTTPClient *http.Client
+	log        *zerolog.Logger
+}
+
+func NewTeslaAPIService(settings *config.Settings, logger *zerolog.Logger) TeslaAPIService {
+	return &teslaAPIService{
+		Settings: settings,
+		HTTPClient: &http.Client{
+			Timeout: 5 * time.Second,
+		},
+		log: logger,
+	}
+}
+
+func (t *teslaAPIService) CompleteTeslaAuthCodeExchange(authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error) {
+	u, err := url.Parse(t.Settings.TeslaAuthorization.TokenAuthURL)
+	if err != nil {
+		return nil, fmt.Errorf("Could not complete tesla auth: %w", err)
+	}
+	f := url.Values{}
+	f.Set("grant_type", "authorization_code")
+	f.Set("client_id", t.Settings.TeslaAuthorization.ClientID)
+	f.Set("client_secret", t.Settings.TeslaAuthorization.ClientSecret)
+	f.Set("code", authCode)
+	f.Set("redirect_uri", redirectURI)
+	f.Set("scope", "openid offline_access "+strings.Join(teslaScopes, " "))
+	f.Set("audience", fmt.Sprintf("https://fleet-api.prd.%s.vn.cloud.tesla.com", region))
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
```

Comment:
Can someone remind me how this interacts with the 5 second timeout that's on the HTTP client we're using? Do we need both?

Note that it's still good to pass in the context so that canceling the request can kill this request to Tesla.

Reply:
The Timout set on the client will override the one used in the context. Personally, I don't love using short Timeouts on clients because it is a hard cap on all requests and encompasses more than just sending the request since it will also close the body if you do not read it before the timeout is exceeded.